From 71f18f6cd6944a9dda41776e9095a553dcb195b9 Mon Sep 17 00:00:00 2001
From: zljing <zljing@codeaurora.org>
Date: Thu, 7 Dec 2017 15:56:55 +0800
Subject: [PATCH 11/14] framework/base: fix multithread synchronization

When do install many apks, null point exception occurs
occasionally due to variable pkg unsynchronized. Add
lock to sync pkg.

CRs-Fixed: 2155091

Change-Id: Ib257bb6e1aa3d78400bc5651c55d0cc9f38375ea
---
 core/java/android/content/pm/PackageParser.java | 28 +++++++++++++------------
 1 file changed, 15 insertions(+), 13 deletions(-)

diff --git a/core/java/android/content/pm/PackageParser.java b/core/java/android/content/pm/PackageParser.java
index bc8ef25..602b7d3 100644
--- a/core/java/android/content/pm/PackageParser.java
+++ b/core/java/android/content/pm/PackageParser.java
@@ -1692,19 +1692,21 @@ public class PackageParser {
                             }
                             final Signature[] entrySignatures = convertToSignatures(entryCerts);
 
-                            if (pkg.mCertificates == null) {
-                                pkg.mCertificates = entryCerts;
-                                pkg.mSignatures = entrySignatures;
-                                pkg.mSigningKeys = new ArraySet<PublicKey>();
-                                for (int i=0; i < entryCerts.length; i++) {
-                                    pkg.mSigningKeys.add(entryCerts[i][0].getPublicKey());
-                                }
-                            } else {
-                                if (!Signature.areExactMatch(pkg.mSignatures, entrySignatures)) {
-                                    throw new PackageParserException(
-                                            INSTALL_PARSE_FAILED_INCONSISTENT_CERTIFICATES, "Package " + apkPath
-                                            + " has mismatched certificates at entry "
-                                            + entry.getName());
+                            synchronized (sObjWaitAll) {
+                                if (pkg.mCertificates == null) {
+                                    pkg.mCertificates = entryCerts;
+                                    pkg.mSignatures = entrySignatures;
+                                    pkg.mSigningKeys = new ArraySet<PublicKey>();
+                                    for (int i=0; i < entryCerts.length; i++) {
+                                        pkg.mSigningKeys.add(entryCerts[i][0].getPublicKey());
+                                    }
+                                } else {
+                                    if (!Signature.areExactMatch(pkg.mSignatures, entrySignatures)) {
+                                        throw new PackageParserException(
+                                                INSTALL_PARSE_FAILED_INCONSISTENT_CERTIFICATES, "Package " + apkPath
+                                                + " has mismatched certificates at entry "
+                                                + entry.getName());
+                                    }
                                 }
                             }
                         } catch (GeneralSecurityException e) {
-- 
2.7.4

